keps/sig-node/5307-container-restart-policy/README.md

```diff
@@ -0,0 +1,1017 @@
+ignored on kubelet and container needs to be restarted quickly.
+
+See https://github.com/kubernetes-sigs/jobset/issues/876 for the detailed description.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+The container level restart policy may restart the container differently from
+the existing assumptions on pod level restart policy. Pod with `restartPolicy=Never`
+may have containers restarted and have the restart count higher than 0. This may
+also affect how Job `podFailurePolicy` interacts with pod failures.
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+#### Controllers managing pod failures
+
+Controllers that manage workloads based on Pod failures (like Job with
+podFailurePolicy) might misinterpret the Pod's state. A container restarting
+due to an "ignored" exit code might not increment the Pod's restart count
+in a way that these controllers expect for failure accounting.
+
+In this case, Job controller with podFailurePolicy only execute the policy
+after the Pod is no longer in Running phase. A container restart will not
+change the Pod's phase. Therefore this works fine with Job's podFailurePolicy
+without any necessary changes to the Job controller or Job API.
+
+#### Unintended Restart Loops
+
+A container might persistently exit with an "ignored" exit code due to
+an unresolvable underlying issue, leading to frequent restarts that consume
+node resources and potentially mask the problem.
+
+The container restart will still follow the exponential backoff to avoid
+excessive resource consumption due to restarts.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The proposal is to implement a simple API with the very limited set of
+allowed values to [Container](https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L2528)
+under k8s.io/apis/core. The shape of API is informed by some future improvements
+we may need to implement as described here:
+https://github.com/kubernetes/enhancements/issues/3329#issuecomment-1571643421
+
+```yaml
+restartPolicy: Never
+containers:
+- name: myapp1
+  # the default behavior is inherited from the Pod’s restartPolicy
+  restartPolicy: Custom 
```

Comment:
I would suggest to consider alternative names, maybe:
- "RuleBased" 
- "Conditional"

I'm leaning to "RuleBased" to match the "restartRules" name.